File: 오픈채팅방.py
def solution(record):
    answer = []
    nickname_by_id = {}
    for re in record:
        commands = re.split(" ")
        state = commands[0]
        
        if state == "Enter" or state == "Change":
            user_id = commands[1]
            user_nickname = commands[2]
            nickname_by_id[user_id] = user_nickname
        
    
    for re in record:
        commands = re.split(" ")
        state = commands[0]
        if state == "Enter":
            user_id = commands[1]
            user_nickname = commands[2]
            phrase = nickname_by_id[user_id]+"님이 들어왔습니다."
            answer.append(phrase)
        elif state == "Leave":
            user_id = commands[1]
            phrase = nickname_by_id[user_id]+"님이 나갔습니다."
            answer.append(phrase)
    return answer

File: java/오픈채팅방.java
//03:17~ 4:23 집중!
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        
        List<Map<String,String>> results = new ArrayList<>();
        Map<String, String> namesById = new HashMap<>();
        
        int count = 0;
        
        for(String re: record){
            String[] test = re.split(" ");
            String state = test[0];
            // System.out.println("state:" + state);
            if (state.equals("Enter")){
                String id = test[1];
                String name = test[2];
                namesById.put(id,name);
                count+=1;
            }else if (state.equals("Leave")){
                String id = test[1];
                count+=1;
            }else if (state.equals("Change")){
                String id = test[1];
                String name = test[2];
                namesById.put(id,name);
            }
        }
        String[] answer = new String[count];
        
        int ct=0;
        for(String re: record){
            String[] test = re.split(" ");
            String state = test[0];
            
            if (state.equals("Enter")){
                String id = test[1];
                String name = test[2];
                answer[ct++] = namesById.get(id) + "님이 들어왔습니다.";
            }else if (state.equals("Leave")){
                String id = test[1];
                answer[ct++] = namesById.get(id) + "님이 나갔습니다.";
            }
        }
        
        return answer;
    }
}